README.md
# :unicorn: Qontalk: A Go SDK for Qontak

![License](https://img.shields.io/badge/license-MIT-blue.svg)
![Go version](https://img.shields.io/badge/go-1.16%2B-blue.svg)
[![Go Reference](https://pkg.go.dev/badge/github.com/maskentir/qontalk.svg)](https://pkg.go.dev/github.com/maskentir/qontalk)

Qontalk is a Go SDK that provides provides a unified Go SDK for seamless interaction with both the Qontak and FSM APIs, allowing you to interact with the Qontak platform. This SDK includes functionalities to authenticate, send message interactions, send interactive messages, send WhatsApp messages, send Direct WhatsApp Broadcasts, and retrieve WhatsApp Templates.

For more detailed information on using the Qontak Go SDK, check out the [GODOC.md](GODOC.md) file.

### Overview :bulb:

The qontalk package combines the functionality of the Qontak and FSM packages to provide a single, comprehensive SDK for both Qontak messaging and Finite State Machine \(FSM\) operations. This package allows you to effortlessly work with Qontak for messaging while simultaneously building, managing, and executing FSMs within your applications.

### Qontak Integration :cactus:

The Qontak integration within qontalk enables you to:

- Authenticate with Qontak's services using various authentication methods.
- Send messages to customers and agents, including text, images, and interactive messages.
- Manage WhatsApp templates and send WhatsApp messages.
- Perform Direct WhatsApp Broadcasts with customization options.

You can utilize these features to enhance your messaging capabilities and communication with customers and agents through Qontak's platform.

### Finite State Machine \(FSM\) Integration :rocket:

The FSM integration allows you to create, manage, and execute Finite State Machines within your application. You can define custom states, events, transitions, and callbacks to control the flow of your application based on specific conditions.

Key features of the FSM integration include:

- Creating FSM instances with custom initial states and transitions.
- Sending events to trigger state transitions.
- Managing FSM lifecycle, including starting and stopping FSM execution.
- Defining global callbacks to handle state transitions and events.

This FSM integration empowers you to build complex, stateful applications with ease.

## Installation

To install the Qontalk SDK, you can use `go get`:

```sh
go get github.com/maskentir/qontalk
```

## Documentation

You can find the full documentation for the Qontalk SDK [here](https://pkg.go.dev/github.com/maskentir/qontalk).

## Usage

Here is a simple example of how to use the Qontalk SDK:

```go
package main

import (
    "fmt"
    "github.com/maskentir/qontalk"
    "github.com/maskentir/qontalk/fsm"
)

func main() {
    // Create a QontalkSDK instance
    sdk := qontalk.NewQontalkSDKBuilder().
        WithClientCredentials("your-username", "your-password", "your-grant-type", "your-client-id", "your-client-secret").
        Build()

    // Authenticate with Qontak
    if err := sdk.Authenticate(); err != nil {
        fmt.Println("Authentication failed:", err)
        return
    }

    // Use Qontak features, send messages, etc.

    // Create an FSM instance
    fsm := fsm.NewBot("ChatBot")

    fsm.AddState("start", "Hi there! Reply with one of the following options:\n1 View growth history\n2 Update growth data\nExample: type '1' if you want to view your child's growth history.", []fsm.Transition{
        {Event: "1", Target: "view_growth_history"},
        {Event: "2", Target: "update_growth_data"},
    }, []fsm.Rule{}, fsm.Rule{})

    fsm.AddState("view_growth_history", "Growth history of your child: Name: {{child_name}} Height: {{height}} Weight: {{weight}} Month: {{month}}", []fsm.Transition{
        {Event: "exit", Target: "start"},
    }, []fsm.Rule{}, fsm.Rule{
        Name:    "custom_error",
        Pattern: regexp.MustCompile("error"),
        Respond: "Custom error message for view_growth_history state.",
    })

    fsm.AddState("update_growth_data", "Please provide the growth information for your child. Use this template e.g., 'Month: January Child's name: John Weight: 30.5 kg Height: 89.1 cm'", []fsm.Transition{
        {Event: "exit", Target: "start"},
    }, []fsm.Rule{}, fsm.Rule{
        Name:    "custom_error",
        Pattern: regexp.MustCompile("error"),
        Respond: "Custom error message for update_growth_data state.",
    })

    fsm.AddRuleToState("update_growth_data", "rule_update_growth_data", `Month: (?P<month>.+) Child's name: (?P<child_name>.+) Weight: (?P<weight>.+) kg Height: (?P<height>.+) cm`, "Thank you for updating {{child_name}}'s growth in {{month}} with height {{height}} and weight {{weight}}", nil)

    messages := []string{
        "2",
        "Month: January Child's name: John Weight: 30.5 kg Height: 89.1 cm",
        "error",
    }

    for _, message := range messages {
        response, err := fsm.ProcessMessage("user1", message)
        if err != nil {
            fmt.Printf("Error processing message '%s': %v\n", message, err)
        } else {
            fmt.Printf("User1: %s\n", message)
            fmt.Printf("Bot: %s\n", response)
        }
    }
}
```

## Contributing

Please read our [Contribution Guidelines](CONTRIBUTE.md) before submitting a pull request.

## Code of Conduct

Please follow our [Code of Conduct](CODE_OF_CONDUCT.md) when participating in this project.

## Issue and Pull Request Templates

Before creating an issue or pull request, review our [Issue Template](ISSUE_TEMPLATE.md) and [Pull Request Template](PULL_REQUEST_TEMPLATE.md).

## License

This library is released under the MIT [License](LICENSE).

## Contact

If you have any questions or feedback, please contact our support team at harunwols@gmail.com.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

### Added

- New feature or enhancement.

### Changed

- Changes to existing functionality.

### Fixed

- Bug fixes or issue resolutions.

## [1.2.2] - 2023-09-02

### Added

- **FSM Integration**: Integrated a powerful Finite State Machine (FSM) library, enabling efficient management of complex state transitions for chatbot development. 🤖🔄

### Fixed

- Resolved issues related to FSM implementation and interactions. 🐛
- Improved error handling and overall performance. 🛠️

## [1.2.1] - 2023-08-29

### Added

- **New MD Files**: We've added new Markdown files to enhance project documentation. 📝
- **Enhanced Contribution Guidelines**: Updated contribution guidelines for a smoother collaboration process. 🤝
- **Improved Code of Conduct**: Our Code of Conduct has been updated for a more inclusive community. 🌈

### Enhanced

- **Improved README**: We've made the README more informative and user-friendly. 📚
- **Updated Issue and Pull Request Templates**: Enhanced templates for a better issue and pull request experience. 📄
- **Refined Code Style**: Improved coding style for consistency. 🎨

### Fixed

- **Fixed Codeowner File**: The CODEOWNER file has been corrected. 🧐
- **Updated Release Guidelines**: Updated release documentation for clarity. 📦

## [1.2.0] - 2023-08-29

### Added

- Qontak SDK Integration: The Qontalk SDK is now seamlessly integrated into the Qontak ecosystem, allowing you to interact with Qontak's messaging and communication services effortlessly. 📞💬
- Finite State Machine (FSM) Support: The SDK includes a powerful Finite State Machine (FSM) library, making it easier than ever to manage complex state transitions in your applications. 🔄
- WhatsApp Integration: Send WhatsApp messages directly from your Go applications using the Qontalk Go SDK. 📱💬
- Direct WhatsApp Broadcast: Send direct WhatsApp broadcasts with custom parameters, including message templates, language settings, and buttons. 📢
- Improved Documentation: Our documentation has been enhanced to provide more detailed information and examples, making it easier for developers to get started. 📖

### Fixed

- Fixed several issues related to message interactions and interactive messages. 🐛
- Improved error handling and error messages for a smoother development experience. 🛠️
- Enhanced performance and reliability in various scenarios. 💪

## [1.1.0] - 2023-08-15

### Changed

- Introduced PutMultipart magic to MockRequestStrategy
- Revamped unit test for SendInteractiveMessage to embrace PutMultipart wizardry
- Nurtured test coverage and code tidying
- Sprinkled some developer love and bug fixes

## [1.0.0] - 2023-08-14

### Added

- Initial implementation of the project.

[Unreleased]: https://github.com/maskentir/qontalk/compare/v1.2.1...HEAD
[1.2.2]: https://github.com/maskentir/qontalk/releases/tag/v1.2.2
[1.2.1]: https://github.com/maskentir/qontalk/releases/tag/v1.2.1
[1.2.0]: https://github.com/maskentir/qontalk/releases/tag/v1.2.0
[1.1.0]: https://github.com/maskentir/qontalk/releases/tag/v1.1.0
[1.0.0]: https://github.com/maskentir/qontalk/releases/tag/v1.0.0


doc.go
// Package qontalk provides a unified Go SDK for seamless interaction with both the Qontak and FSM APIs.
//
// # Overview
//
// The qontalk package combines the functionality of the Qontak and FSM packages
// to provide a single, comprehensive SDK for both Qontak messaging and Finite State
// Machine (FSM) operations. This package allows you to effortlessly work with Qontak
// for messaging while simultaneously building, managing, and executing FSMs within
// your applications.
//
// # Qontak Integration
//
// The Qontak integration within qontalk enables you to:
//
// - Authenticate with Qontak's services using various authentication methods.
// - Send messages to customers and agents, including text, images, and interactive messages.
// - Manage WhatsApp templates and send WhatsApp messages.
// - Perform Direct WhatsApp Broadcasts with customization options.
//
// You can utilize these features to enhance your messaging capabilities and
// communication with customers and agents through Qontak's platform.
//
// # Finite State Machine (FSM) Integration
//
// The FSM integration allows you to create, manage, and execute Finite State Machines
// within your application. You can define custom states, events, transitions, and
// callbacks to control the flow of your application based on specific conditions.
//
// Key features of the FSM integration include:
//
// - Creating FSM instances with custom initial states and transitions.
// - Sending events to trigger state transitions.
// - Managing FSM lifecycle, including starting and stopping FSM execution.
// - Defining global callbacks to handle state transitions and events.
//
// This FSM integration empowers you to build complex, stateful applications with ease.
//
// # Example Usage
//
// The following example demonstrates how to leverage the qontalk package to interact
// with both Qontak and FSM functionalities:
//
//	package main
//
//	import (
//	    "fmt"
//	    "github.com/maskentir/qontalk"
//	    "github.com/maskentir/qontalk/fsm"
//	)
//
//	func main() {
//	    // Create a QontalkSDK instance
//	    sdk := qontalk.NewQontalkSDKBuilder().
//	        WithClientCredentials("your-username", "your-password", "your-grant-type", "your-client-id", "your-client-secret").
//	        Build()
//
//	    // Authenticate with Qontak
//	    if err := sdk.Authenticate(); err != nil {
//	        fmt.Println("Authentication failed:", err)
//	        return
//	    }
//
//	    // Use Qontak features, send messages, etc.
//
//	    // Create an FSM instance
//	    fsm := fsm.NewBot("ChatBot")
//
//	    fsm.AddState("start", "Hi there! Reply with one of the following options:\n1 View growth history\n2 Update growth data\nExample: type '1' if you want to view your child's growth history.", []fsm.Transition{
//	        {Event: "1", Target: "view_growth_history"},
//	        {Event: "2", Target: "update_growth_data"},
//	    }, []fsm.Rule{}, fsm.Rule{})
//
//	    fsm.AddState("view_growth_history", "Growth history of your child: Name: {{child_name}} Height: {{height}} Weight: {{weight}} Month: {{month}}", []fsm.Transition{
//	        {Event: "exit", Target: "start"},
//	    }, []fsm.Rule{}, fsm.Rule{
//	        Name:    "custom_error",
//	        Pattern: regexp.MustCompile("error"),
//	        Respond: "Custom error message for view_growth_history state.",
//	    })
//
//	    fsm.AddState("update_growth_data", "Please provide the growth information for your child. Use this template e.g., 'Month: January Child's name: John Weight: 30.5 kg Height: 89.1 cm'", []fsm.Transition{
//	        {Event: "exit", Target: "start"},
//	    }, []fsm.Rule{}, fsm.Rule{
//	        Name:    "custom_error",
//	        Pattern: regexp.MustCompile("error"),
//	        Respond: "Custom error message for update_growth_data state.",
//	    })
//
//	    fsm.AddRuleToState("update_growth_data", "rule_update_growth_data", `Month: (?P<month>.+) Child's name: (?P<child_name>.+) Weight: (?P<weight>.+) kg Height: (?P<height>.+) cm`, "Thank you for updating {{child_name}}'s growth in {{month}} with height {{height}} and weight {{weight}}", nil)
//
//	    messages := []string{
//	        "2",
//	        "Month: January Child's name: John Weight: 30.5 kg Height: 89.1 cm",
//	        "error",
//	    }
//
//	    for _, message := range messages {
//	        response, err := fsm.ProcessMessage("user1", message)
//	        if err != nil {
//	            fmt.Printf("Error processing message '%s': %v\n", message, err)
//	        } else {
//	            fmt.Printf("User1: %s\n", message)
//	            fmt.Printf("Bot: %s\n", response)
//	        }
//	    }
//	}
//
// This example showcases how you can use the qontalk package to work with Qontak
// messaging and FSM features in a single application, creating a unified experience.
//
// # Additional Resources
//
// For more detailed documentation and comprehensive usage examples, please refer to
// the individual package documentation for qontak and fsm.
package qontalk
